lib/puppet/provider/nameservice.rb

```diff
@@ -55,7 +55,7 @@ def listbyname
       Etc.send("set#{section()}ent")
       begin
         while ent = Etc.send("get#{section()}ent")
-          names << ent.name
+          names << ent.name.force_encoding(Encoding::UTF_8)
```

Comment:
I think we want `encode` here, rather than `force_encoding`, maybe falling back to `force_encoding` if for some reason the local locale (heh) has characters that can't be represented in Unicode. `force_encoding` just reinterprets the bytes as the new encoding, which turns them in to garbage - they may not even be (and probably aren't, if we're coming from a different encoding) valid UTF-8 data.

`encode` instead will do the proper conversion, taking each character from the external encoding to utf-8.